### DISCUSSION TITLE
Debugging the BlackHole: Virtual Audio Driver

### DISCUSSION THREAD

**Discussion by Raj123456788** [ExistentialAudio/BlackHole#248] [General]
Can you please update the steps to debug this HAL plugin?

**Comment by devinroth** [comment]
Just print messages to the console.

**Reply by Raj123456788** [reply]
Sorry, I am new to the Xcode. Can you please jot down basic steps will be great? Where will it print the messages in log? Also, did you use any template to get started? Thanks!

**Comment by devinroth** [comment]
What are you trying to do?

Devin

**Reply by Raj123456788** [reply]
Trying to understand how it has been written and understanding HAL Plugin(since Kernel Drivers will be soon obsoleted especially with M1 chipset macs)  https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/CoreAudioOverview/CoreAudioEssentials/CoreAudioEssentials.html.

**Comment by devinroth** [comment]
The debug messages are sent to the console.

Not sure what else to add. If you have any specific questions let me know.

Devin

**Reply by Raj123456788** [reply]
Which console?
I am able to build the plugin and copy to /Library/Audio/Plug-Ins. it works but where does it logs messages? Any tips to share that were helpful during developing this HAL Plugin?

**Comment by devinroth** [comment]
Console is an Mac App that system logs all the system logs. All debug messages can be found there. 

Devin

**Reply by Raj123456788** [reply]
Do I need to create the mac app or is there an example/sample app that I can use? Can you please provide details.

**Comment by devinroth** [comment]
Just do a spotlight search for console. You’ll find it. 

Devin

**Reply by Raj123456788** [reply]
Hello again, is there a template that was used to write this driver? Please post the important links that were useful during the development. A good tutorial would be really helpful. Thanks!

**Comment by devinroth** [comment]
If you are interested in developing your own, you are on your own. All the information is available on the web and through reading the API/docs/headers. I am not interested in holding your hand through the process.

Devin
